src/source/Ice/Tls.c

```diff
@@ -0,0 +1,267 @@
+/**
+ * Kinesis Video Tcp
+ */
+#define LOG_CLASS "Tls"
+#include "../Include_i.h"
+
+STATUS createTlsSession(TlsSessionCallbacks callbacks, PTlsSession *ppTlsSession)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PTlsSession pTlsSession = NULL;
+
+    CHK(ppTlsSession != NULL, STATUS_NULL_ARG);
+    CHK(callbacks.outboundPacketFn != NULL, STATUS_NULL_ARG);
+
+    pTlsSession = MEMCALLOC(1, SIZEOF(TlsSession));
+    CHK(pTlsSession != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    pTlsSession->callbacks = callbacks;
+    pTlsSession->state = TLS_SESSION_STATE_NEW;
+
+CleanUp:
+    if (STATUS_FAILED(retStatus) && pTlsSession != NULL) {
+        freeTlsSession(&pTlsSession);
+    }
+
+    if (ppTlsSession != NULL) {
+        *ppTlsSession = pTlsSession;
+    }
+    
+    LEAVES();
+    return retStatus;
+}
+
+STATUS freeTlsSession(PTlsSession* ppTlsSession)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PTlsSession pTlsSession = NULL;
+
+    CHK(ppTlsSession != NULL, STATUS_NULL_ARG);
+
+    pTlsSession = *ppTlsSession;
+    CHK(pTlsSession != NULL, retStatus);
+
+    if (pTlsSession->pSslCtx != NULL) {
+        SSL_CTX_free(pTlsSession->pSslCtx);
+    }
+
+    if (pTlsSession->pSsl != NULL) {
+        SSL_free(pTlsSession->pSsl);
+    }
+
+    CHK_STATUS(tlsSessionShutdown(pTlsSession));
```

Comment:
Can't do CHK_STATUS here. In case of an error it will skip the mem freeing a line below. Store the return value perhaps and return it later